cluster-autoscaler/processors/provreq/injector.go

```diff
@@ -44,57 +45,85 @@ type ProvisioningRequestPodsInjector struct {
+func (p *ProvisioningRequestPodsInjector) FilterProcess(
+	filter func(*provreqwrapper.ProvisioningRequest) bool,
 ) ([]*apiv1.Pod, error) {
 	provReqs, err := p.client.ProvisioningRequests()
 	if err != nil {
 		return nil, err
 	}
+
 	for _, pr := range provReqs {
-		if ok, found := provisioningrequest.SupportedProvisioningClasses[pr.Spec.ProvisioningClassName]; !ok || !found {
-			klog.Warningf("Provisioning Class %s is not supported", pr.Spec.ProvisioningClassName)
-			continue
-		}
-		conditions := pr.Status.Conditions
-		if apimeta.IsStatusConditionTrue(conditions, v1.Failed) || apimeta.IsStatusConditionTrue(conditions, v1.Provisioned) {
+		if !filter(pr) {
 			continue
 		}
 
-		provisioned := apimeta.FindStatusCondition(conditions, v1.Provisioned)
-
 		//TODO(yaroslava): support exponential backoff
 		// Inject pods if ProvReq wasn't scaled up before or it has Provisioned == False condition more than defaultRetryTime
-		inject := true
-		if provisioned != nil {
-			if provisioned.Status == metav1.ConditionFalse && provisioned.LastTransitionTime.Add(defaultRetryTime).Before(p.clock.Now()) {
-				inject = true
-			} else {
-				inject = false
-			}
+		if !p.IsAvailableForProvisioning(pr) {
+			continue
 		}
-		if inject {
-			provreqpods, err := provreqpods.PodsForProvisioningRequest(pr)
-			if err != nil {
-				klog.Errorf("Failed to get pods for ProvisioningRequest %v", pr.Name)
-				provreqconditions.AddOrUpdateCondition(pr, v1.Failed, metav1.ConditionTrue, provreqconditions.FailedToCreatePodsReason, err.Error(), metav1.NewTime(p.clock.Now()))
-				if _, err := p.client.UpdateProvisioningRequest(pr.ProvisioningRequest); err != nil {
-					klog.Errorf("failed add Failed condition to ProvReq %s/%s, err: %v", pr.Namespace, pr.Name, err)
-				}
-				continue
-			}
-			provreqconditions.AddOrUpdateCondition(pr, v1.Accepted, metav1.ConditionTrue, provreqconditions.AcceptedReason, provreqconditions.AcceptedMsg, metav1.NewTime(p.clock.Now()))
-			if _, err := p.client.UpdateProvisioningRequest(pr.ProvisioningRequest); err != nil {
-				klog.Errorf("failed add Accepted condition to ProvReq %s/%s, err: %v", pr.Namespace, pr.Name, err)
-				continue
-			}
-			unschedulablePods := append(unschedulablePods, provreqpods...)
-			return unschedulablePods, nil
+
+		podsFromProvReq, err := provreqpods.PodsForProvisioningRequest(pr)
+		if err != nil {
+			klog.Errorf("Failed to get pods for ProvisioningRequest %v", pr.Name)
+			p.MarkAsFailed(pr, err.Error())
+			continue
+		}
+
+		if err := p.MarkAsAccepted(pr); err != nil {
+			continue
 		}
+		return podsFromProvReq, nil
 	}
-	return unschedulablePods, nil
+	return nil, nil
+}
+
+// Process pick one ProvisioningRequest, update Accepted condition and inject pods to unscheduled pods list.
+func (p *ProvisioningRequestPodsInjector) Process(
+	_ *context.AutoscalingContext,
+	unschedulablePods []*apiv1.Pod,
+) ([]*apiv1.Pod, error) {
+	return p.FilterProcess(
```

Comment:
This will replace current unschedulablePods with those from ProvisioningRequest, but what we did here before was append the new pods to unschedulablePods. 

If we replace the pods, it'll effectively disable scale-ups for unschedulable pods in the cluster (i.e. break autoscaling cluster-wide).

Reply:
updated to append the pods